include/particles/twophase_particle.tcc

```diff
@@ -0,0 +1,737 @@
+//! Construct a two phase particle with id and coordinates
+template <unsigned Tdim>
+mpm::TwoPhaseParticle<Tdim>::TwoPhaseParticle(Index id, const VectorDim& coord)
+    : mpm::Particle<Tdim>(id, coord) {
+  this->initialise();
+  // Clear cell ptr
+  cell_ = nullptr;
+  // Nodes
+  nodes_.clear();
+  // Set material containers
+  this->initialise_material(2);
+  // Logger
+  std::string logger =
+      "twophaseparticle" + std::to_string(Tdim) + "d::" + std::to_string(id);
+  console_ = std::make_unique<spdlog::logger>(logger, mpm::stdout_sink);
+}
+
+//! Construct a twophase particle with id, coordinates and status
+template <unsigned Tdim>
+mpm::TwoPhaseParticle<Tdim>::TwoPhaseParticle(Index id, const VectorDim& coord,
+                                              bool status)
+    : mpm::Particle<Tdim>(id, coord, status) {
+  this->initialise();
+  cell_ = nullptr;
+  nodes_.clear();
+  // Set material containers
+  this->initialise_material(2);
+  //! Logger
+  std::string logger =
+      "twophaseparticle" + std::to_string(Tdim) + "d::" + std::to_string(id);
+  console_ = std::make_unique<spdlog::logger>(logger, mpm::stdout_sink);
+}
+
+//! Return particle data in HDF5 format
+template <unsigned Tdim>
+// cppcheck-suppress *
+mpm::HDF5Particle mpm::TwoPhaseParticle<Tdim>::hdf5() const {
+  // Derive from particle
+  auto particle_data = mpm::Particle<Tdim>::hdf5();
+  // // Particle liquid velocity
+  // Eigen::Vector3d liquid_velocity;
+  // liquid_velocity.setZero();
+  // for (unsigned j = 0; j < Tdim; ++j)
+  //   liquid_velocity[j] = this->liquid_velocity_[j];
+  // // Particle liquid mass
+  // particle_data.liquid_mass = this->liquid_mass_;
+  // Particle pore pressure
+  // particle_data.pressure = this->pore_pressure_;
+  // // Particle liquid velocity
+  // particle_data.liquid_velocity_x = liquid_velocity[0];
+  // particle_data.liquid_velocity_y = liquid_velocity[1];
+  // particle_data.liquid_velocity_z = liquid_velocity[2];
+  // // Particle liquid material id
+  // particle_data.liquid_material_id = this->liquid_material_id_;
+  if (this->material(mpm::ParticlePhase::Liquid) != nullptr) {
+    particle_data.nstate_vars =
+        state_variables_[mpm::ParticlePhase::Liquid].size();
+    if (state_variables_[mpm::ParticlePhase::Liquid].size() > 20)
+      throw std::runtime_error("# of state variables cannot be more than 20");
+    unsigned i = 0;
+    auto state_variables =
+        (this->material(mpm::ParticlePhase::Liquid))->state_variables();
+    for (const auto& state_var : state_variables) {
+      particle_data.svars[i] =
+          state_variables_[mpm::ParticlePhase::Liquid].at(state_var);
+      ++i;
+    }
+  }
+
+  return particle_data;
+}
```

Comment:
This is for the output variables. We have now few additional data that we want to print as well as the second `state_vars` for the `Liquid` phase.